contracts/DaiHrd.sol

```diff
@@ -464,4 +504,99 @@ contract DaiHrd is ERC777 {
+		uint256 attodaiHrd = balanceOf(tokenHolder);
+		return convertAttodaiHrdToAttodai(attodaiHrd, rontodaiPerPot);
+	}
+
+	function totalSupplyDai() external view returns(uint256 attodai) {
+		uint256 rontodaiPerPot = calculatedChi();
+		attodai = convertAttodaiHrdToAttodai(totalSupply(), rontodaiPerPot);
+		return attodai;
+	}
+
+	function withdrawDai(uint256 attodai) external returns(uint256 attodaiHrd) {
+		uint256 rontodaiPerPot = updateAndFetchChi();
+		attodaiHrd = convertAttodaiToAttodaiHrd(attodai, rontodaiPerPot);
+		uint256 attodaiWithdrawn = _withdraw(attodaiHrd);
+		require(attodaiWithdrawn > attodai);
+		return attodaiHrd;
+	}
+
+	function sendDenominatedInDai(address recipient, uint256 attodai, bytes calldata data) external {
+		uint256 rontodaiPerPot = calculatedChi();
+		uint256 attodaiHrd = convertAttodaiToAttodaiHrd(attodai, rontodaiPerPot);
+		_send(_msgSender(), _msgSender(), recipient, attodaiHrd, data, "", true);
+	}
+
+	function burnDenominatedInDai(uint256 attodai, bytes calldata data) external {
+		uint256 rontodaiPerPot = calculatedChi();
+		uint256 attodaiHrd = convertAttodaiToAttodaiHrd(attodai, rontodaiPerPot);
+		_burn(_msgSender(), _msgSender(), attodaiHrd, data, "");
+	}
+
+	function operatorSendDenominatedInDai(address sender, address recipient, uint256 attodai, bytes calldata data, bytes calldata operatorData) external {
+		require(isOperatorFor(_msgSender(), sender), "ERC777: caller is not an operator for holder");
+
+		uint256 rontodaiPerPot = calculatedChi();
+		uint256 attodaiHrd = convertAttodaiToAttodaiHrd(attodai, rontodaiPerPot);
+		_send(_msgSender(), sender, recipient, attodaiHrd, data, operatorData, true);
+	}
+
+	function operatorBurnDenominatedInDai(address account, uint256 attodai, bytes calldata data, bytes calldata operatorData) external {
+		require(isOperatorFor(_msgSender(), account), "ERC777: caller is not an operator for holder");
+
+		uint256 rontodaiPerPot = calculatedChi();
+		uint256 attodaiHrd = convertAttodaiToAttodaiHrd(attodai, rontodaiPerPot);
+		_burn(_msgSender(), account, attodaiHrd, data, operatorData);
+	}
+
+	// Utility Functions
+	function convertAttodaiToAttodaiHrd(uint256 attodai, uint256 rontodaiPerPot ) internal pure returns (uint256 attodaiHrd) {
+		// + 1 is to compensate rounding, since attodaiHrd is rounded down
+		attodaiHrd = attodai.mul(ONE).div(rontodaiPerPot) + 1;
+		return attodaiHrd;
+	}
+
+	function convertAttodaiHrdToAttodai(uint256 attodaiHrd, uint256 rontodaiPerPot ) internal pure returns (uint256 attodai) {
+		attodai = attodaiHrd.mul(rontodaiPerPot).div(ONE);
+		return attodai;
+	}
+
+	function calculatedChi() internal view returns (uint256 rontodaiPerPot) {
+		rontodaiPerPot = rmul(rpow(pot.dsr(), now - pot.rho(), ONE), pot.chi());
+		return rontodaiPerPot;
+	}
+
+	function updateAndFetchChi() internal returns (uint256 rontodaiPerPot) {
+		if (pot.rho() == now) {
+			rontodaiPerPot = pot.chi(); // replace with drop()?? TODO
```

Comment:
It is rapidly looking less and less likely that `drop()` will make it in.  We may want to create a `public` helper method for:
```solidity
if (pot.rho() == now) return pot.drip();
else return pot.chi();
```
just so we don't have to duplicate that code (we do that in at least a few places).

Reply:
Oh, 🤣 this _is_ that helper method!